File: nim_core_v2/ios/Classes/services/FLTConversationIdUtil.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation
import NIMSDK

enum ConversationIdUtilType: String {
  case conversationId
  case p2pConversationId
  case teamConversationId
  case superTeamConversationId
  case conversationType
  case conversationTargetId
  case isConversationIdValid
  case sessionTypeV1
}

class FLTConversationIdUtil: FLTBaseService, FLTService {
  func serviceName() -> String {
    ServiceType.ConversationIdUtil.rawValue
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any], _ resultCallback: ResultCallback) {
    switch method {
    case ConversationIdUtilType.p2pConversationId.rawValue:
      p2pConversationId(arguments, resultCallback)
    case ConversationIdUtilType.teamConversationId.rawValue:
      teamConversationId(arguments, resultCallback)
    case ConversationIdUtilType.superTeamConversationId.rawValue:
      superTeamConversationId(arguments, resultCallback)
    case ConversationIdUtilType.conversationType.rawValue:
      conversationType(arguments, resultCallback)
    case ConversationIdUtilType.conversationTargetId.rawValue:
      conversationTargetId(arguments, resultCallback)
    default:
      resultCallback.notImplemented()
    }
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  func conversationId(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let conversationId = arguments["conversationId"] as? String, let conversationTypeInt = arguments["conversationType"] as? Int, let conversationType = V2NIMConversationType(rawValue: conversationTypeInt) else {
      parameterError(resultCallback)
      return
    }
  }

  func p2pConversationId(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountId = arguments["accountId"] as? String else {
      parameterError(resultCallback)
      return
    }
    let covnersationId = V2NIMConversationIdUtil.p2pConversationId(accountId)
    successCallBack(resultCallback, covnersationId)
  }

  func teamConversationId(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let teamId = arguments["teamId"] as? String else {
      parameterError(resultCallback)
      return
    }
    let covnersationId = V2NIMConversationIdUtil.teamConversationId(teamId)
    successCallBack(resultCallback, covnersationId)
  }

  func superTeamConversationId(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let superTeamId = arguments["superTeamId"] as? String else {
      parameterError(resultCallback)
      return
    }
    let covnersationId = V2NIMConversationIdUtil.superTeamConversationId(superTeamId)
    successCallBack(resultCallback, covnersationId)
  }

  func conversationType(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let conversationId = arguments["conversationId"] as? String else {
      parameterError(resultCallback)
      return
    }
    let covnersationType = V2NIMConversationIdUtil.conversationType(conversationId)
    successCallBack(resultCallback, ["conversationType": covnersationType.rawValue])
  }

  func conversationTargetId(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let conversationId = arguments["conversationId"] as? String else {
      parameterError(resultCallback)
      return
    }
    let covnersationTargetId = V2NIMConversationIdUtil.conversationTargetId(conversationId)
    successCallBack(resultCallback, covnersationTargetId)
  }

  func isConversationIdValid(_ arguments: [String: Any], _ resultCallback: ResultCallback) {}

  func sessionTypeV1(_ arguments: [String: Any], _ resultCallback: ResultCallback) {}
}

File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/services/FLTConversationIdUtil.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import android.text.TextUtils
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimlib.sdk.msg.constant.SessionTypeEnum
import com.netease.nimlib.sdk.v2.conversation.enums.V2NIMConversationType
import com.netease.nimlib.sdk.v2.utils.V2NIMConversationIdUtil

class FLTConversationIdUtil(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {
    private val tag = "FLTConversationIdUtil"
    override val serviceName = "ConversationIdUtil"

    init {
        registerFlutterMethodCalls(
            "conversationId" to ::conversationId,
            "p2pConversationId" to ::p2pConversationId,
            "teamConversationId" to ::teamConversationId,
            "superTeamConversationId" to ::superTeamConversationId,
            "conversationType" to ::conversationType,
            "conversationTargetId" to ::conversationTargetId,
            "isConversationIdValid" to ::isConversationIdValid,
            "sessionTypeV1" to ::sessionTypeV1
        )
    }

    private suspend fun conversationId(
        arguments: Map<String, *>
    ): NimResult<String> {
        val targetId = arguments["targetId"] as? String
        val type =
            (arguments["conversationType"] as? Int)?.let {
                V2NIMConversationType.typeOfValue(it)
            }

        val sessionId = arguments["sessionId"] as? String
        val sessionType =
            (arguments["sessionType"] as? Int)?.let {
                SessionTypeEnum.typeOfValue(it)
            }
        return if (!TextUtils.isEmpty(targetId) && type != null) {
            NimResult(code = 0, data = V2NIMConversationIdUtil.conversationId(targetId, type))
        } else if (!TextUtils.isEmpty(sessionId) && sessionType != null) {
            NimResult(code = 0, data = V2NIMConversationIdUtil.conversationId(sessionId, sessionType))
        } else {
            NimResult(code = -1, data = "Invalid arguments")
        }
    }

    private suspend fun p2pConversationId(
        arguments: Map<String, *>
    ): NimResult<String> {
        val accountId = arguments["accountId"] as? String
        val conversationId = V2NIMConversationIdUtil.p2pConversationId(accountId)
        return NimResult(code = 0, data = conversationId)
    }

    private suspend fun teamConversationId(
        arguments: Map<String, *>
    ): NimResult<String> {
        val teamId = arguments["teamId"] as? String
        return NimResult(code = 0, data = V2NIMConversationIdUtil.teamConversationId(teamId))
    }

    private suspend fun superTeamConversationId(
        arguments: Map<String, *>
    ): NimResult<String> {
        val superTeamId = arguments["superTeamId"] as? String
        return NimResult(code = 0, data = V2NIMConversationIdUtil.superTeamConversationId(superTeamId))
    }

    private suspend fun conversationType(
        arguments: Map<String, *>
    ): NimResult<Map<String, Any?>?> {
        val conversationId = arguments["conversationId"] as? String
        val type = V2NIMConversationIdUtil.conversationType(conversationId)
        return NimResult(code = 0, data = mapOf("conversationType" to type.value))
    }

    private suspend fun conversationTargetId(
        arguments: Map<String, *>
    ): NimResult<String> {
        val conversationId = arguments["conversationId"] as? String
        val targetId = V2NIMConversationIdUtil.conversationTargetId(conversationId)
        return NimResult(code = 0, data = targetId)
    }

    private suspend fun isConversationIdValid(
        arguments: Map<String, *>
    ): NimResult<Boolean> {
        val conversationId = arguments["conversationId"] as? String
        return NimResult(code = 0, data = V2NIMConversationIdUtil.isConversationIdValid(conversationId))
    }

    private suspend fun sessionTypeV1(
        arguments: Map<String, *>
    ): NimResult<Map<String, Any?>?> {
        val type =
            (arguments["conversationType"] as? Int)?.let {
                V2NIMConversationType.typeOfValue(it)
            }
        val sessionType = V2NIMConversationIdUtil.sessionTypeV1(type)
        return NimResult(code = 0, data = mapOf("SessionType" to sessionType?.value))
    }
}
